Game/Plugins/MetricsServiceProvider/Source/MetricsServiceProvider/Private/MetricsServiceModule.cpp

```diff
@@ -0,0 +1,617 @@
+	Attributes.Add(GameCurrencyAmountAttr);
+
+	RecordEvent(TEXT("CurrencyPurchased"), Attributes);
+	UE_LOG(LogAnalytics, Display, TEXT("Metrics::RecordCurrencyPurchase('%s', %d, %d)"), *GameCurrencyType, GameCurrencyAmount, Attributes.Num());
+}
+
+void FAnalyticsProviderMetrics::RecordCurrencyGiven(const FString& GameCurrencyType, int GameCurrencyAmount, const TArray<FAnalyticsEventAttribute>& EventAttrs)
+{
+	TArray<FAnalyticsEventAttribute> Attributes;
+	const FAnalyticsEventAttribute GameCurrencyTypeAttr = FAnalyticsEventAttribute(TEXT("GameCurrencyType"), GameCurrencyType);
+	const FAnalyticsEventAttribute GameCurrencyAmountAttr = FAnalyticsEventAttribute(TEXT("GameCurrencyAmount"), FString::FromInt(GameCurrencyAmount));
+
+	Attributes.Append(EventAttrs);
+	Attributes.Add(GameCurrencyTypeAttr);
+	Attributes.Add(GameCurrencyAmountAttr);
+
+	RecordEvent(TEXT("CurrencyGiven"), Attributes);
+	UE_LOG(LogAnalytics, Display, TEXT("Metrics::RecordCurrencyGiven('%s', %d, %d)"), *GameCurrencyType, GameCurrencyAmount, Attributes.Num());
+}
+
+void FAnalyticsProviderMetrics::RecordError(const FString& Error, const TArray<FAnalyticsEventAttribute>& EventAttrs)
+{
+	TArray<FAnalyticsEventAttribute> Attributes;
+	const FAnalyticsEventAttribute GameErrorAttr = FAnalyticsEventAttribute(TEXT("Error"), Error);
+
+	Attributes.Append(EventAttrs);
+	Attributes.Add(GameErrorAttr);
+
+	RecordEvent(TEXT("Error"), Attributes);
+	UE_LOG(LogAnalytics, Display, TEXT("Metrics::RecordError('%s', %d)"), *Error, EventAttrs.Num());
+}
+
+void FAnalyticsProviderMetrics::RecordProgress(const FString& ProgressType, const FString& ProgressHierarchy, const TArray<FAnalyticsEventAttribute>& EventAttrs)
+{
+	TArray<FAnalyticsEventAttribute> Attributes;
+	const FAnalyticsEventAttribute ProgressTypeAttr = FAnalyticsEventAttribute(TEXT("ProgressType"), ProgressType);
+	const FAnalyticsEventAttribute ProgressHierarchyAttr = FAnalyticsEventAttribute(TEXT("ProgressHierarchy"), ProgressHierarchy);
+
+	Attributes.Append(EventAttrs);
+	Attributes.Add(ProgressTypeAttr);
+	Attributes.Add(ProgressHierarchyAttr);
+
+	RecordEvent(TEXT("Progress"), Attributes);
+	UE_LOG(LogAnalytics, Display, TEXT("Metrics::RecordProgress('%s', %s, %d)"), *ProgressType, *ProgressHierarchy, Attributes.Num());
+}
+
+FString FAnalyticsProviderMetrics::MetricsPayload::Stringify() const
+{
+	TSharedRef<FJsonObject> Obj = MakeShared<FJsonObject>();
+	TSharedRef<FJsonObject> Attributes = MakeShared<FJsonObject>();
+
+	Obj->SetStringField(TEXT("eventEnvironment"), EventEnvironment);
+	Obj->SetStringField(TEXT("eventSource"), EventSource);
+	Obj->SetStringField(TEXT("eventClass"), EventClass);
+	Obj->SetStringField(TEXT("eventType"), EventType);
+	Obj->SetNumberField(TEXT("eventTimestamp"), EventTimestamp);
+	Obj->SetNumberField(TEXT("eventIndex"), EventIndex);
+
+	Obj->SetStringField(TEXT("versionId"), EngineVersion);
+
+	auto EmptyCheckToSetField = [Obj](const FString& CheckID, const FString& SetField) {
```

Comment:
Hmm should this not be &Obj?

Reply:
Obj is a pointer, if not change the pointer, directly pass a pointer will be better than pass by reference

Reply:
Ahh yes it is a pointer... silly me...